File: mvp/src/main/java/com/thefuntasty/taste/injection/module/BaseActivityModule.java
package com.thefuntasty.taste.injection.module;

import android.app.Activity;
import android.content.Context;

import com.thefuntasty.taste.TasteActivity;
import com.thefuntasty.taste.injection.annotation.qualifier.ActivityContext;
import com.thefuntasty.taste.injection.annotation.scope.PerScreen;

import dagger.Module;
import dagger.Provides;

@Module
public abstract class BaseActivityModule<T extends TasteActivity> {

	private final T activity;

	public BaseActivityModule(T activity) {
		this.activity = activity;
	}

	@Provides
	@PerScreen
	public T activityT() {
		return activity;
	}

	@Provides
	@PerScreen
	public Activity activity() {
		return activity;
	}

	@Provides
	@ActivityContext
	public Context context() {
		return activity;
	}
}

File: mvp/src/main/java/com/thefuntasty/taste/injection/module/BaseActivityModule.kt
package com.thefuntasty.taste.injection.module

import android.app.Activity
import android.arch.lifecycle.Lifecycle
import android.content.Context

import com.thefuntasty.taste.TasteActivity
import com.thefuntasty.taste.injection.annotation.ActivityContext
import com.thefuntasty.taste.injection.annotation.PerScreen

import dagger.Module
import dagger.Provides

@Module
abstract class BaseActivityModule<out T : TasteActivity>(private val activity: T) {

	@Provides
	@PerScreen
	fun activityT(): T {
		return activity
	}

	@Provides
	@PerScreen
	fun activity(): Activity {
		return activity
	}

	@Provides
	@ActivityContext
	fun context(): Context {
		return activity
	}

    @Provides
    @ActivityContext
    fun lifecycle(): Lifecycle {
        return activity.lifecycle
    }
}
